### PR TITLE
Add RunCommand functions to support vVols

### PR DESCRIPTION
**Pull request by hfanan**
This PR closes #

The changes in this PR are as follows:
Add functions to:
- create and remove datastores
- register VASA provider
- add storage policies

Testing: Unit tests using Pester running on on-premise VMWare instances

I have read the [contributor guidelines](CONTRIBUTING.md) and have completed the following:

* [x] **Formatted the code** using VSCode default formatter for PowerShell.
* [ ] **Tested the code** end-to-end against an SDDC. 
          >  Not possible before deploying to Azure.
* [x] **Documented the functions** using standard PowerShell markup and applied `AVSAttribute` to newly exported functions.

### CODE REVIEW

**Review by et1975 — COMMENTED** [review]
I'm assuming the intention is to use this from your automation? If yes let's mark the cmdlets for automation only, for example
`[AVSAttribute(10, UpdatesSDDC = $false, AutomationOnly = $true)]`

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psd1 [hunk 1]
```diff
@@ -0,0 +1,136 @@
+#
+# Module manifest for module 'Microsoft.AVS.VMFS'
```
[the remark this answers is not in the document]
Comment (hfanan) [reply Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psd1:2]: Good catch. Will update

**Comment by hfanan** [comment]
> [AVSAttribute(10, UpdatesSDDC = $false, AutomationOnly = $true)]

I am not familiar with the AutomationOnly attribute. What is the effect of adding it?

**Comment by et1975** [comment]
> @et1975 : I am not familiar with the AutomationOnly attribute. What is the effect of adding it?

It's exactly what it says on the can - the script won't show up to a human looking at the list in the portal, but will be still available for execution via the API.

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psd1 [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+    # Minimum version of the common language runtime (CLR) required by this module. This prerequisite is valid for the PowerShell Desktop edition only.
+    # ClrVersion = ''
+
+    # Processor architecture (None, X86, Amd64) required by this module
+    # ProcessorArchitecture = ''
+
+    # Modules that must be imported into the global environment prior to importing this module
+    RequiredModules   = @(
+        @{ "ModuleName" = "Microsoft.AVS.Management"; "ModuleVersion" = "5.2.39" }
```
Comment (et1975) [inline Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psd1:51]: We're on https://www.powershellgallery.com/packages/Microsoft.AVS.Management/5.3.88 at the moment, let's have that version as the dependency.

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psd1 [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        "Remove-VvolVasaProvider",
+        "New-VvolStoragePolicy",
+        "Remove-VvolStoragePolicy"
+    )
+
+    # Cmdlets to export from this module, for best performance, do not use wildcards and do not delete the entry, use an empty array if there are no cmdlets to export.
+    CmdletsToExport   = @()
+
+    # Variables to export from this module
+    VariablesToExport = '*'
```
Comment (et1975) [inline Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psd1:84]: Do we need any variables exported? Didn't seem so.

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+    .EXAMPLE
+     Remove-VVolDatastore -ClusterName "myCluster" -DatastoreName "myDatastore"
+
+    .INPUTS
+     vCenter cluster name, datastore name
+
+    .OUTPUTS
+     None.
+#>
+function Remove-VvolDatastore {
```
Comment (et1975) [inline Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1:94]: Could we check as a precondition that there are no VMs using this datastore?

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+    .EXAMPLE
+     Remove-VvolVasaProvider -ProviderName "myProvider"
+
+    .INPUTS
+     vCenter VASA provider name
+
+    .OUTPUTS
+     None.
+#>
+function Remove-VvolVasaProvider {
```
Comment (et1975) [inline Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1:239]: Ditto - can we check that the provider is not in use. If not directly, then by looking at the datastores it provides?

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+}
+
+
+<#
+    .SYNOPSIS
+     Creates a new vVol Storage Policy
+
+    .PARAMETER PolicyConfigJsonString
+     Storage policy in Json string. The command will traverse through the SPBM rules and creat a storage policy containing all of the rules.
+     The example string is as below:
```
Comment (et1975) [inline Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1:268]: Let's not specify the example in the param description. Please move it to `.EXAMPLE`.

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+            }
+            if ($ValueType -eq "TimeSpan") {
+                $RuleValue = [TimeSpan]::Parse($RuleValue)
+            }
+        }
+        $rules += New-SpbmRule -Capability (Get-SpbmCapability -Name $RuleName) -Value $RuleValue
+    }
+
+    Write-Host "Creating policy $PolicyName for vendor $Vendor..."
+    $Ruleset = New-SpbmRuleSet -AllOfRules $rules
```
Comment (et1975) [inline Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1:331]: How do we make sure that the policy is only applicable to VVOL objects?

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+    [AVSAttribute(10, UpdatesSDDC = $false)]
+    Param(
+        [Parameter(
+            Mandatory = $true,
+            HelpMessage = 'Storage policy name')]
+        [ValidateNotNull()]
+        [String]$PolicyName
+    )
+
+    $Policy = Get-SpbmStoragePolicy -Name $PolicyName -ErrorAction Ignore
```
Comment (et1975) [inline Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1:362]: Ditto.

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+            Mandatory = $true,
+            HelpMessage = 'Name of VASA provider to be created in vCenter')]
+        [ValidateNotNull()]
+        [String]
+        $ProviderName
+    )
+
+    $VasaProvider = Get-VasaProvider -Name $ProviderName -ErrorAction Ignore
+    if (-not $VasaProvider) {
+        throw "Unable to remove a VASA provider. Vasa provider '$ProviderName' does not exist."
```
Comment (srajmohan24) [inline Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1:253]: Nit: VASA provider '$ProviderName'

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+            HelpMessage = 'Storage policy in Json string')]
+        [ValidateNotNull()]
+        [String]$PolicyConfigJsonString
+    )
+    # Convert JSON to PowerShell object
+    $PolicyConfig = ConvertFrom-Json $PolicyConfigJsonString
+    # Extract values from the object
+    $Vendor = $PolicyConfig.Vendor
+    $PolicyName = $PolicyConfig.PolicyName
+    $PolicyDescription = $PolicyConfig.PolicyDescription
```
Comment (srajmohan24) [inline Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1:306]: Could we have a check at the beginning of this cmdlet to ensure that $Vendor, $PolicyName and $PolicyDescription values are not empty?

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psd1 (hunk 2 above)
Comment (hfanan) [reply Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psd1]: Updated

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psd1 (hunk 3 above)
Comment (hfanan) [reply Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psd1]: We do not need it. I removed it

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 (hunk 4 above)
Comment (hfanan) [reply Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1:94]: Done

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 (hunk 6 above)
Comment (hfanan) [reply Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1]: Done

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 (hunk 5 above)
Comment (hfanan) [reply Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1:243]: Done

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 (hunk 7 above)
Comment (hfanan) [reply Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1:340]: We cannot do this a t the time of creating the policy.

**Comment by hfanan** [comment]
> 

It is valuable for us to show the RunCommand in the GUI as this can help in troubleshooting. If we mark the RunCommands for AutomationOnly, is there a workaround later to enable showing it in the GUI without code update?

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 (hunk 10 above)
Comment (hfanan) [reply Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1:318]: Added check for PolicyName
Vendor and description are optional

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 (hunk 9 above)
Comment (hfanan) [reply Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1]: done

File: Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1 (hunk 8 above)
Comment (hfanan) [reply Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1:374]: Same we can not check which objects are using the policy at the time it is created

**Review by et1975 — APPROVED** [review]
If you intend these cmdlets for the human interaction via portal then sure, keep `AVSAttribute` as is, I assumed it will mirror more or less the VMFS scenario.

**Comment by et1975** [comment]
One thing to keep in mind is that every Microsoft package and cmdlet we show in the Run Command portal may have to be documented on Azure AVS docs before going GA.

**Pull request merged**